File: newproject/store.py
import time 
import qrcode
PRODUCTS = []

def read_from_database():
    f = open('newproject/database.txt','r')
    
    for line in f:
        result = line.split(',')
        my_dict = {"code":result[0],'name':result[1],'price':result[2],'count':result[3]}
        PRODUCTS.append(my_dict)
    f.close() 
    

def show_menu():
    print('1- Add')    
    print('2- Edit')    
    print('3- Remove')    
    print('4- Search')    
    print('5- Show List')    
    print('6- Buy')    
    print('7- product detail QRcode')    
    print('8- Exit')    
      
def make_QRcode():
    user_choose = input('product id => ')
    for product in PRODUCTS:
        if user_choose == product['code']:
            img = qrcode.make(product)
            img.save('code.png')
def add():
    code = input('enter code => ')
    name = input('enter name => ')
    price = input('enter price => ')
    count = input('enter count => ')
    new_product = {'code':code,"name":name,"price":price,'count':count}
    PRODUCTS.append(new_product)
    print(PRODUCTS)
def edit():
    product_id = input("enter product id => ")
    for product in PRODUCTS:
        if product['code'] == product_id:
            product_edit = int(input('edit name => 1 \t edit price => 2 \t edit count => 3 '))
            if product_edit == 1:
                edit_name = input('write new name ')
                finally_name = product["name"] = edit_name
                PRODUCTS.append(finally_name)
                print("name edited successfully")
                break
            elif product_edit == 2:
                edit_price = int(input('enter new price '))
                finally_price = product['price'] = edit_price
                PRODUCTS.append(finally_price)
                print("price edited successfully")
                break
            elif product_edit == 3:
                edit_count = int(input('enter new count '))
                finally_count = product['count'] = edit_count
                PRODUCTS.append(finally_count)
                print("count edited successfully")
                break
            else:
                print('invalid value')
        else:
            print('not found')
    else:
        
        print("invalid input")        
def remove():
    product_delete = input("product id => ")
    for product in PRODUCTS:
        if product["code"] == product_delete:
            PRODUCTS.remove(product)
            print(f"Product with ID {product_delete} has been removed.")
            break
        else:
            print(f'product with this id => {product_delete} not found') 
def search():
    user_input = input("type your keyword => ")
    for product in PRODUCTS:
       if product['code'] == user_input or product['name'] == user_input: 
        print(product['code'], '\t' , product['name'], '\t' , product["price"], '\t')
        break
    else:
           print('not found')
    
def show_list():
    print("code\tname\tprice")
    for product in PRODUCTS:
        print(product['code'], '\t' , product['name'], '\t' , product["price"], '\t')

def buy():
    ...
    
def set_to_database():
    ...

print('welcome my store')
time.sleep(0.5)
print('Loading.') 
time.sleep(0.5)
print('Loading..') 
time.sleep(0.5)
print('Loading...') 
time.sleep(0.5)
read_from_database()
print('data Loaded')
while True:
    show_menu()
    
    user_input = input('What is your choice => ')
    
    try:
        choice = int(user_input)
        
        if choice == 1:
            add()
        elif choice == 2:
            edit()
        elif choice == 3:
            remove()
        elif choice == 4:
            search()
        elif choice == 5:
            show_list()
        elif choice == 6:
            buy()
        elif choice == 7:
            make_QRcode()
        elif choice == 8:
            set_to_database()
        elif choice == 9:
            print("Exiting program.")
            exit(0)
        else:
            print("Invalid number, please try again.")
    
    except ValueError:
        print("Please enter a valid number.")

File: project 7 - Dictionary/php/store.php
<?php


$PRODUCTS = [];

function read_from_database() {
    global $PRODUCTS;
    if (file_exists('database.txt')) {
        $file = fopen('database.txt', 'r');
        while (($line = fgets($file)) !== false) {
            $result = explode(',', trim($line));
            $PRODUCTS[] = [
                "code" => $result[0],
                "name" => $result[1],
                "price" => floatval($result[2]),
                "count" => intval($result[3])
            ];
        }
        fclose($file);
    }
}

function show_menu() {
    echo "\n1- Add\n";
    echo "2- Edit\n";
    echo "3- Remove\n";
    echo "4- Search\n";
    echo "5- Show List\n";
    echo "6- Buy\n";
    echo "7- Product Detail QRcode\n";
    echo "8- Save to Database\n";
    echo "9- Exit\n";
}

function add() {
    global $PRODUCTS;
    
    $code = readline('Enter code => ');
    $name = readline('Enter name => ');
    $price = floatval(readline('Enter price => '));
    $count = intval(readline('Enter count => '));
    
    $PRODUCTS[] = [
        'code' => $code,
        'name' => $name,
        'price' => $price,
        'count' => $count
    ];
    
    echo "Product added successfully.\n";
}

function edit() {
    global $PRODUCTS;
    
    $product_id = readline("Enter product ID => ");
    
    foreach ($PRODUCTS as &$product) {
        if ($product['code'] == $product_id) {
            $product_edit = intval(readline('Edit name => 1, Edit price => 2, Edit count => 3: '));
            switch ($product_edit) {
                case 1:
                    $edit_name = readline('Write new name: ');
                    $product["name"] = $edit_name;
                    break;
                case 2:
                    $edit_price = floatval(readline('Enter new price: '));
                    $product['price'] = $edit_price;
                    break;
                case 3:
                    $edit_count = intval(readline('Enter new count: '));
                    $product['count'] = $edit_count;
                    break;
                default:
                    echo 'Invalid value\n';
                    return;
            }
            echo "Product edited successfully.\n";
            return;
        }
    }
    
    echo "Product not found.\n";
}

function remove() {
    global $PRODUCTS;

    $product_delete = readline("Product ID => ");
    
    foreach ($PRODUCTS as $key => $product) {
        if ($product["code"] == $product_delete) {
            unset($PRODUCTS[$key]);
            echo "Product with ID {$product_delete} has been removed.\n";
            return;
        }
    }
    
    echo "Product with this ID {$product_delete} not found.\n"; 
}

function search() {
    global $PRODUCTS;

    $user_input = readline("Type your keyword => ");
    
    foreach ($PRODUCTS as $product) {
        if ($product['code'] == $user_input || stripos($product['name'], $user_input) !== false) { 
            echo "{$product['code']}\t{$product['name']}\t{$product['price']}\t{$product['count']}\n";
            return;
        }
    }
    
    echo 'Not found\n';
}

function show_list() {
    global $PRODUCTS;

    echo "Code\tName\tPrice\tCount\n";
    
    foreach ($PRODUCTS as $product) {
        echo "{$product['code']}\t{$product['name']}\t{$product['price']}\t{$product['count']}\n";
    }
}

function buy() {
    global $PRODUCTS;

    $user_input = readline("Enter the Product ID you want to buy: ");
    $quantity = intval(readline("Enter quantity: "));
    
    foreach ($PRODUCTS as &$product) {
        if ($product['code'] == $user_input) {
            if ($quantity <= $product['count']) {
                $total_cost = number_format($quantity * $product['price'], 2);
                $product['count'] -= $quantity;
                echo "You bought {$quantity} of {$product['name']} for \${$total_cost}.\n";

                if ($product['count'] == 0) {
                    remove();
                }
                return;
            } else {
                echo "Insufficient stock. Available count is {$product['count']}.\n";
                return;
            }
        }
    }

    echo "Product not found.\n";
}

function make_QRcode() {
    global $PRODUCTS;

    require_once 'phpqrcode/qrlib.php';

    $user_choose = readline('Product ID => ');
    
    foreach ($PRODUCTS as $product) {
        if ($user_choose == $product['code']) {
            QRcode::png(json_encode($product), 'code.png');
            echo "QR code generated and saved as 'code.png'.\n";
            return;
        }
    }

    echo "Product not found.\n";
}

function set_to_database() {
    global $PRODUCTS;

    if (empty($PRODUCTS)) {
        echo "No products to save.\n";
        return;
    }

    file_put_contents('database.txt', '');
    
    foreach ($PRODUCTS as $product) {
        file_put_contents('database.txt', implode(',', array_values($product)) . "\n", FILE_APPEND);
    }

   echo "Data saved to database.\n";
}

read_from_database();
echo "Data Loaded\n";

while (true) {
   show_menu();
   
   try {
       $user_input = readline('What is your choice => ');
       $choice = intval($user_input);
       
       switch ($choice) {
           case 1: add(); break;
           case 2: edit(); break;
           case 3: remove(); break;
           case 4: search(); break;
           case 5: show_list(); break;
           case 6: buy(); break;
           case 7: make_QRcode(); break; 
           case 8: set_to_database(); break;
           case 9:
               echo "Exiting program.\n";
               exit(0);
           default:
               echo "Invalid number, please try again.\n";
               break; 
       }

   } catch (ValueError | Exceptione) { 
       echo "Please enter a valid number.\n"; 
   }
}
?>